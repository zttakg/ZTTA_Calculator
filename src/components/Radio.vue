<template>
  <label :class="classes" :for="id">
    <input
      class="form-control form-control_type_radio radio__control"
      type="radio"
      :id="id"
      :name="name"
      :value="value"
      :class="className"
      :required="required"
      :disabled="disabled"
      @change="onChange"
      :checked="state"
    />
    <span class="radio__box" v-if="boxHelper"></span>
    <span v-if="label || hasSlot()"
      :class="[
        'radio__label',
        { 'visuallyhidden' :labelHidden }
      ]">
      <template v-if="!hasSlot()">{{ label }}</template>
      <slot></slot>
    </span>
  </label>
</template>

<script>
import uid from "@/helpers/generateId";
import SlotMixin from "@/mixins/slot";

export default {
  name: "Radio",

  mixins: [SlotMixin],

  model: {
    prop: "modelValue",
    event: "input"
  },

  props: {
    id: {
      type: String,
      default() {
        return `radio-id-${uid()}`;
      }
    },
    name: {
      type: String,
      default: null
    },
    value: {
      default: ""
    },
    modelValue: {
      default: undefined
    },
    className: {
      type: String,
      default: null
    },
    theme: {
      type: String,
      default: "default"
    },
    label: String,
    checked: {
      type: Boolean,
      default: false
    },
    required: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    labelHidden: {
      type: Boolean,
      default: false
    },
    boxHelper: {
      type: Boolean,
      default: true
    },
    baseClass: {
      type: String,
      default: "radio"
    },
    model: {}
  },

  computed: {
    classes() {
      const themes = this.theme
        .split(" ")
        .filter(t => t !== "")
        .map(t => `${this.baseClass}_style_${t}`);

      return [
        this.baseClass,
        {
          [`${this.baseClass}_checked`]: this.state
        },
        ...themes
      ];
    },
    state() {
      if (this.modelValue === undefined) {
        return this.checked;
      }
      return this.modelValue === this.value;
    }
  },

  watch: {
    checked(newValue) {
      if (newValue !== this.state) {
        this.toggle();
      }
    }
  },

  mounted() {
    if (this.checked && !this.state) {
      this.toggle();
    }
  },

  methods: {
    onChange() {
      this.toggle();
    },
    toggle() {
      this.$emit("input", this.state ? "" : this.value);
    }
  }
};
</script>
