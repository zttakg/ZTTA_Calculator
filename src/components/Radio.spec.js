import { shallowMount } from "@vue/test-utils";

import Radio from "./Radio";

jest.mock("../helpers/generateId", () => () => "mockedUID");

describe("<Radio /> component", () => {
  it("should render correct without props", () => {
    const wrapper = shallowMount(Radio);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render all props", () => {
    const wrapper = shallowMount(Radio, {
      propsData: {
        id: "test-id",
        name: "test-name",
        value: "",
        className: "test-class",
        theme: "test-theme",
        label: "Test label",
        checked: true,
        required: true,
        disabled: true,
        labelHidden: true
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
