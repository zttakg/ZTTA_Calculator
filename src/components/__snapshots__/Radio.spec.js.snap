// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> component should render all props 1`] = `
<label for="test-id" class="radio radio_checked radio_style_test-theme">
  <input type="radio" id="test-id" name="test-name" required="required" disabled="disabled" class="form-control form-control_type_radio radio__control test-class" value=""> <span class="radio__box"></span> <span class="radio__label visuallyhidden">Test label </span></label>
`;

exports[`<Radio /> component should render correct without props 1`] = `
<label for="radio-id-mockedUID" class="radio radio_style_default">
  <input type="radio" id="radio-id-mockedUID" class="form-control form-control_type_radio radio__control" value=""> <span class="radio__box"></span>
  <!---->
</label>
`;
